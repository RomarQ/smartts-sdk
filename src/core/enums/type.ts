export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
