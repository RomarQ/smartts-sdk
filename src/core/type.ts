import { IToString } from '../typings';

export class BaseType {
    private _ = 'Type';
}

export class SimpleType extends BaseType {
    constructor(public name: string) {
        super();
    }

    toString() {
        return `"${this.name}"`;
    }
}

export class T_TList extends BaseType {
    constructor(public innerType: IToString) {
        super();
    }

    toString() {
        return `(list ${this.innerType})`;
    }
}

export const TList = (innerType: BaseType) => new T_TList(innerType);

export const TNat = new SimpleType('nat');
export const TInt = new SimpleType('int');
export const TMutez = new SimpleType('mutez');
export const TString = new SimpleType('string');
