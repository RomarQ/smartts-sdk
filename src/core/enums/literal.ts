export enum LiteralAtom {
    record = 'record',
    variant = 'variant',
    chain_id_cst = 'chain_id_cst',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    Some = 'Some',
    None = 'None',
    Left = 'Left',
    Right = 'Right',
    map = 'map',
    big_map = 'big_map',
    signature = 'signature',
    tuple = 'tuple',
    ticket = 'ticket',
    sapling_state = 'sapling_empty_state',
}

export default LiteralAtom;
