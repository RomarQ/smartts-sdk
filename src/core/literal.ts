import type { IType } from '../typings/type';
import type { ILiteral } from '../typings/literal';

import { capitalizeBoolean, LineInfo } from '../misc/utils';
import {
    TNat,
    TString,
    TMutez,
    TAddress,
    TBool,
    TList,
    TTimestamp,
    TChainID,
    TInt,
    TBytes,
    TOption,
    TUnknown,
    TUnit,
    TRecord,
} from './type';
import { Prim } from './enums/prim';
import { IExpressionKind } from '../typings/expression';
import { Layout } from './enums/layout';

class Literal implements ILiteral {
    _isLiteral = true as const;

    constructor(
        public name: string,
        public value: number | string | boolean | undefined,
        public type: IType,
        public line: LineInfo,
    ) {}

    toString() {
        if (typeof this.value === 'undefined') {
            return `(${this.name})`;
        }
        return `(literal (${this.name} ${this.value}) ${this.line})`;
    }

    toType() {
        return this.type.toString();
    }
}

class ListLiteral implements ILiteral {
    _isLiteral = true as const;

    constructor(public name: string, public items: IExpressionKind[], public type: IType, public line: LineInfo) {}

    toString() {
        return `(${this.name} ${this.line}  ${this.items.map((item) => item.toString()).join(' ')})`;
    }

    toType() {
        return this.type.toString();
    }
}

class OptionLiteral implements ILiteral {
    _isLiteral = true as const;

    type: IType;

    constructor(
        public prim: Prim.Some | Prim.None,
        public value: IExpressionKind | undefined,
        public innerType: IType = TUnknown,
        public line: LineInfo,
    ) {
        this.type = TOption(innerType);
    }

    toString(): string {
        return `(variant "${this.prim}" ${this.value?.toString() || '(unit)'} ${this.line})`;
    }

    toType() {
        return this.type.toString();
    }
}

class RecordLiteral implements ILiteral {
    _isLiteral = true as const;

    type: IType;

    constructor(public fields: Record<string, ILiteral>, public line: LineInfo) {
        // Compute the record type (use rightcombs by default)
        this.type = TRecord(
            Object.entries(fields).reduce(
                (pv, [field, expr]) => ({
                    ...pv,
                    [field]: expr.type.toString(),
                }),
                {},
            ),
            Layout.right_comb,
        );
    }

    static buildFields = (fields: Record<string, IExpressionKind>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(record ${this.line} ${RecordLiteral.buildFields(this.fields).join(' ')})`;
    }

    toType() {
        return this.type.toString();
    }
}

export const Unit = (line = new LineInfo()) => new Literal(Prim.unit, undefined, TUnit, line);

export const Nat = (value: number, line = new LineInfo()) => new Literal(Prim.nat, value, TNat, line);
export const Int = (value: number, line = new LineInfo()) => new Literal(Prim.int, value, TInt, line);
export const Mutez = (value: number, line = new LineInfo()) => new Literal(Prim.mutez, value, TMutez, line);

export const String = (value: string, line = new LineInfo()) => new Literal(Prim.string, `"${value}"`, TString, line);

export const Bool = (value: boolean, line = new LineInfo()) =>
    new Literal(Prim.bool, capitalizeBoolean(value), TBool, line);

export const Address = (address: string, line = new LineInfo()) => new Literal(Prim.address, address, TAddress, line);

export const Timestamp = (timestamp: number, line = new LineInfo()) =>
    new Literal(Prim.timestamp, timestamp, TTimestamp, line);

export const ChainID = (chainID: string, line = new LineInfo()) => new Literal('chain_id_cst', chainID, TChainID, line);

export const Bytes = (bytes: string, line = new LineInfo()) => new Literal(Prim.bytes, bytes, TBytes, line);

export const List = (items: IExpressionKind[], innerType: IType, line = new LineInfo()) =>
    new ListLiteral(Prim.list, items, TList(innerType), line);

export const Some = (value: IExpressionKind, innerType?: IType, line = new LineInfo()) =>
    new OptionLiteral(Prim.Some, value, innerType, line);

export const None = (innerType?: IType, line = new LineInfo()) =>
    new OptionLiteral(Prim.None, undefined, innerType, line);

export const Record = (fields: Record<string, ILiteral>, line = new LineInfo()) => new RecordLiteral(fields, line);

const Literals = {
    Unit,
    Nat,
    Int,
    Mutez,
    String,
    Bool,
    Address,
    Timestamp,
    ChainID,
    Bytes,
    List,
    Some,
    None,
    Record,
};

export default Literals;
