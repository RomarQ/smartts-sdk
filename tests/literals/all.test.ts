import {
    TAddress,
    TBls12_381_fr,
    TBls12_381_g1,
    TBls12_381_g2,
    TBool,
    TBytes,
    TChain_id,
    TInt,
    TKey,
    TKey_hash,
    TMutez,
    TNat,
    TSignature,
    TString,
    TTimestamp,
    TUnit,
} from '../../src/core/type';
import {
    Int,
    Mutez,
    Nat,
    Unit,
    String,
    Bool,
    Bytes,
    Address,
    Timestamp,
    Chain_id,
    Bls12_381_fr,
    Bls12_381_g2,
    Bls12_381_g1,
    Key_hash,
    Signature,
    Key,
} from '../../src/core/literal';
import { ILiteral } from '../../src/typings/literal';
import { IType } from '../../src/typings/type';
import { verifyMichelsonOutput } from '../util';
import { Contract } from '../../src';

const verifyLiteral = (testName: string, type: IType, literal: ILiteral<unknown>) => {
    it(testName, () => {
        const contract = new Contract().setStorageType(type).setStorage(literal).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
};

describe('Test Literals', () => {
    verifyLiteral('unit', TUnit(), Unit());
    verifyLiteral('nat', TNat(), Nat(1));
    verifyLiteral('int', TInt(), Int(2));
    verifyLiteral('mutez', TMutez(), Mutez(2));
    verifyLiteral('string', TString(), String('TEST STRING'));
    verifyLiteral('bool', TBool(), Bool(true));
    verifyLiteral('bytes', TBytes(), Bytes('0x00'));
    verifyLiteral('address', TAddress(), Address('tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN'));
    verifyLiteral('timestamp', TTimestamp(), Timestamp(100));
    verifyLiteral('chain_id', TChain_id(), Chain_id('0x00'));
    verifyLiteral('bls12_381_fr', TBls12_381_fr(), Bls12_381_fr('0x01'));
    verifyLiteral('bls12_381_fr', TBls12_381_fr(), Bls12_381_fr(1));
    verifyLiteral('bls12_381_g1', TBls12_381_g1(), Bls12_381_g1('0x00'));
    verifyLiteral('bls12_381_g2', TBls12_381_g2(), Bls12_381_g2('0x00'));
    verifyLiteral('key', TKey(), Key('edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT'));
    verifyLiteral('key_hash', TKey_hash(), Key_hash('tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN'));
    verifyLiteral(
        'signature',
        TSignature(),
        Signature('sigsAujsNePapNNGsVotTvcKWMNNJja9B4a2FfAe8vExzFhEgEo1GTQStiif22uSA6iNxPGCGsXsRyeLHzeLbJL2y8CnYuNe'),
    );
});
