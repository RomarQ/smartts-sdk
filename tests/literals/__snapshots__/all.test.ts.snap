// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Literals address 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (address tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"all.test.ts\\" 55))
            storage_type (\\"address\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals address 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"address\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bls12_381_fr 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bls12_381_fr 0x01) (\\"all.test.ts\\" 55))
            storage_type (\\"bls12_381_fr\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bls12_381_fr 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bls12_381_fr\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bls12_381_fr 3`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bls12_381_fr 1) (\\"all.test.ts\\" 55))
            storage_type (\\"bls12_381_fr\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bls12_381_fr 4`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bls12_381_fr\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bls12_381_g1 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bls12_381_g1 0x00) (\\"all.test.ts\\" 55))
            storage_type (\\"bls12_381_g1\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bls12_381_g1 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bls12_381_g1\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bls12_381_g2 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bls12_381_g2 0x00) (\\"all.test.ts\\" 55))
            storage_type (\\"bls12_381_g2\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bls12_381_g2 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bls12_381_g2\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bool 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bool True) (\\"all.test.ts\\" 55))
            storage_type (\\"bool\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bool 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bool\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals bytes 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (bytes 0x00) (\\"all.test.ts\\" 55))
            storage_type (\\"bytes\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals bytes 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bytes\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals chain_id 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (chain_id_cst 0x00) (\\"all.test.ts\\" 55))
            storage_type (\\"chain_id\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals chain_id 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"chain_id\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals int 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (int 2) (\\"all.test.ts\\" 55))
            storage_type (\\"int\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals int 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"int\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals key 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (key edpku3g7CeTEvSKhxipD4Q2B6EiEP8cR323u8PFmGFgKRVRvCneEmT) (\\"all.test.ts\\" 55))
            storage_type (\\"key\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals key 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"key\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals key_hash 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (key_hash tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"all.test.ts\\" 55))
            storage_type (\\"key_hash\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals key_hash 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"key_hash\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals list 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (list (\\"all.test.ts\\" 55) (literal (string \\"HELLO\\") (\\"all.test.ts\\" 55)) (literal (string \\"WORLD\\") (\\"all.test.ts\\" 55)))
            storage_type ((list \\"string\\"))
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals list 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"list\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"string\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals mutez 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (mutez 2) (\\"all.test.ts\\" 55))
            storage_type (\\"mutez\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals mutez 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"mutez\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals nat 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (nat 1) (\\"all.test.ts\\" 55))
            storage_type (\\"nat\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals nat 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"nat\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals set 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (set (\\"all.test.ts\\" 55) (literal (string \\"HELLO\\") (\\"all.test.ts\\" 55)) (literal (string \\"WORLD\\") (\\"all.test.ts\\" 55)))
            storage_type ((set \\"string\\"))
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals set 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"set\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"string\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals signature 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (signature sigsAujsNePapNNGsVotTvcKWMNNJja9B4a2FfAe8vExzFhEgEo1GTQStiif22uSA6iNxPGCGsXsRyeLHzeLbJL2y8CnYuNe) (\\"all.test.ts\\" 55))
            storage_type (\\"signature\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals signature 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"signature\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals string 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (string \\"TEST STRING\\") (\\"all.test.ts\\" 55))
            storage_type (\\"string\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals string 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"string\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals timestamp 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (literal (timestamp 100) (\\"all.test.ts\\" 55))
            storage_type (\\"timestamp\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals timestamp 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"timestamp\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Test Literals unit 1`] = `
"
        (
            template_id (static_id 0 (\\"all.test.ts\\" 48))
            storage (unit)
            storage_type (\\"unit\\")
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"all.test.ts\\" 48))
        )
        "
`;

exports[`Test Literals unit 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;
